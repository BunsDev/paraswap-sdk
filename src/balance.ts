import { Token, Address, constructToken, AddressOrSymbol } from './token';
import {
  ConstructFetchInput,
  TokenApiResponse,
  TokensApiResponse,
} from './types';

type GetBalances = (userAddress: Address) => Promise<Token[]>;
type GetBalance = (
  userAddress: Address,
  tokenAddressOrSymbol: AddressOrSymbol
) => Promise<Token | typeof NOT_FOUND_RESPONSE>;

export interface Allowance {
  tokenAddress: Address;
  allowance: string;
}

type GetAllowances = (
  userAddress: Address,
  tokenAddresses: Address[]
) => Promise<Allowance[]>;
type GetAllowance = (
  userAddress: Address,
  tokenAddress: Address
) => Promise<Allowance | typeof NOT_FOUND_RESPONSE>;

export type GetBalancesFunctions = {
  getBalance: GetBalance;
  getBalances: GetBalances;
  getAllowance: GetAllowance;
  getAllowances: GetAllowances;
};

const NOT_FOUND_RESPONSE = { message: 'Not Found' } as const;

export const constructGetBalances = ({
  apiURL,
  network,
  fetcher,
}: ConstructFetchInput): GetBalancesFunctions => {
  const tokensUrl = `${apiURL}/users/tokens/${network}`;

  const getBalances: GetBalances = async (userAddress) => {
    const fetchURL = `${tokensUrl}/${userAddress}`;

    const data = await fetcher<TokensApiResponse>({
      url: fetchURL,
      method: 'GET',
    });

    const tokens = data.tokens.map(constructToken);

    return tokens;
  };

  const getBalance: GetBalance = async (userAddress, tokenAddressOrSymbol) => {
    const fetchURL = `${tokensUrl}/${userAddress}/${tokenAddressOrSymbol}`;

    const data = await fetcher<TokenApiResponse>({
      url: fetchURL,
      method: 'GET',
    });

    if (!data.token) {
      return NOT_FOUND_RESPONSE;
    }

    return constructToken(data.token);
  };

  const getAllowances: GetAllowances = async (userAddress, tokenAddresses) => {
    const tokenAddressesLowercaseSet = new Set(
      tokenAddresses.map((address) => address.toLowerCase())
    );

    const balances = await getBalances(userAddress);

    const allowances: Allowance[] = balances
      .filter((token) =>
        tokenAddressesLowercaseSet.has(token.address.toLowerCase())
      )
      .map(({ address, allowance = '0' }) => ({
        tokenAddress: address,
        allowance,
      }));

    return allowances;
  };

  const getAllowance: GetAllowance = async (userAddress, tokenAddress) => {
    const balanceOrNotFound = await getBalance(userAddress, tokenAddress);

    if ('message' in balanceOrNotFound) {
      return balanceOrNotFound;
    }

    const { address, allowance = '0' } = balanceOrNotFound;

    return { tokenAddress: address, allowance };
  };

  return {
    getBalance,
    getBalances,
    getAllowance,
    getAllowances,
  };
};
