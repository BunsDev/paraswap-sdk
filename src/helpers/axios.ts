import { FetcherFunction } from "../types";
import type Axios from "axios";
import type { AxiosError } from "axios";

type FetcherErrorConstructorInput = Pick<
  AxiosError,
  "code" | "request" | "response" | "isAxiosError" | "message"
>;

interface FetcherErrorInterface extends FetcherErrorConstructorInput {
  status?: number;
  code?: string;
}

export const isFetcherError = (error: unknown): error is FetcherError => {
  return error instanceof FetcherError;
};

export class FetcherError extends Error implements FetcherErrorInterface {
  code;
  status;
  request;
  response;
  isAxiosError = false;

  public constructor({
    code,
    request,
    response,
    isAxiosError,
    message,
  }: FetcherErrorConstructorInput) {
    super();

    // Maintains proper stack trace for where our error was thrown (only available on V8)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, FetcherError);
    }

    this.name = this.constructor.name;
    this.code = code;
    this.request = request;
    this.isAxiosError = isAxiosError;

    // no response on AxiosError
    if (!response) {
      this.message = message;
      return;
    }

    this.response = response;

    const { data, status } = response;
    this.status = status;
    this.message = data?.error || message;
  }
}

export const constructFetcher =
  (axios: typeof Axios): FetcherFunction =>
  async (params) => {
    try {
      const { data } = await axios.request(params);

      return data;
    } catch (error: AxiosError | any) {
      if (axios.isAxiosError(error)) {
        // attach a more meaningful message and other props
        throw new FetcherError(error);
      }

      throw error;
    }
  };
