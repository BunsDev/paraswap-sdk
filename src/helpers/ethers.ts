import type {
  Address,
  ContractCallerFunction,
  NoExtraKeysCheck,
} from '../types';
import { JsonRpcProvider, BaseProvider } from '@ethersproject/providers';
import { Signer } from '@ethersproject/abstract-signer';
import {
  Contract,
  PayableOverrides,
  CallOverrides,
} from '@ethersproject/contracts';
import type { TransactionResponse } from '@ethersproject/abstract-provider';
import { assertContractHasMethods } from './misc';
import { assert } from 'ts-essentials';

export const constructContractCaller = (
  ethers: BaseProvider | Signer,
  account?: Address
): ContractCallerFunction => {
  const contractCallerFunction: ContractCallerFunction = async (params) => {
    if (params.static) {
      const { address, abi, contractMethod, args, overrides } = params;

      const contract = new Contract(address, abi, ethers);

      assertContractHasMethods(contract, contractMethod);
      // drop keys not in CallOverrides
      const { block, gas, ...restOverrides } = overrides;
      // reassign values to keys in CallOverrides
      const normalizedOverrides = {
        ...restOverrides,
        blockTag: block,
        gasLimit: gas,
      };

      // type FinalCallOverrides = normalizedOverrides has extra props ? never : normalizedOverrides
      type FinalCallOverrides = NoExtraKeysCheck<
        typeof normalizedOverrides,
        CallOverrides
      >;

      // enforce overrides shape ethers accepts
      // TS will break if normalizedOverrides type has any keys not also present in CallOverrides
      const callOverrides: FinalCallOverrides = normalizedOverrides;
      // returns whatever the Contract.method returns: BigNumber, string, boolean
      return contract.callStatic[contractMethod](...args, callOverrides);
    }

    assert(account, 'account must be specified to create a signer');
    assert(
      ethers instanceof JsonRpcProvider || ethers instanceof Signer,
      'ethers must be an instance of Signer or JsonRpcProvider to create a signer'
    );

    const { address, abi, contractMethod, args, overrides } = params;

    const signer = 'getSigner' in ethers ? ethers.getSigner(account) : ethers;

    const contract = new Contract(address, abi, signer);

    assertContractHasMethods(contract, contractMethod);
    // drop keys not in PayableOverrides
    const { gas, from, ...restOverrides } = overrides;
    // reassign values to keys in PayableOverrides
    const normalizedOverrides = {
      ...restOverrides,
      gasLimit: gas,
    };

    // type FinalPayableOverrides = normalizedOverrides has extra props ? never : normalizedOverrides
    type FinalPayableOverrides = NoExtraKeysCheck<
      typeof normalizedOverrides,
      PayableOverrides
    >;

    // enforce overrides shape ethers accepts
    // TS will break if normalizedOverrides type has any keys not also present in PayableOverrides
    const txOverrides: FinalPayableOverrides = normalizedOverrides;
    const txResponse: TransactionResponse = await contract[contractMethod](
      ...args,
      txOverrides
    );

    // returns tx hash
    return txResponse.hash;
    // @TODO maybe better return the whole txResponse for versatility
  };

  return contractCallerFunction;
};
