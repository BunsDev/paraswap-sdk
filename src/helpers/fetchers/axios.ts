import type { FetcherFunction } from '../../types';
import type AxiosStatic from 'axios';
import { FetcherError } from '../misc';

export const constructFetcher =
  (axios: typeof AxiosStatic): FetcherFunction =>
  async (params) => {
    try {
      const { data } = await axios.request(params);

      return data;
    } catch (error: any) {
      if (axios.isAxiosError(error)) {
        // attach a more meaningful message and other props
        throw new FetcherError(error);
      }

      throw error;
    }
  };
