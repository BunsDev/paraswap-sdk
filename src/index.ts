import { constructApproveToken, ApproveTokenFunctions } from './approve';
import { constructGetBalances, GetBalancesFunctions } from './balance';
import { constructGetSpender, GetSpenderFunctions } from './spender';
import { constructGetAdapters, AdaptersFunctions } from './adapters';
import { constructGetRate, GetRateFunctions } from './rates';
import {
  constructGetTokens,
  constructToken,
  GetTokensFunctions,
  Token,
  PriceString,
} from './token';
import { BuildTxFunctions, constructBuildTx } from './transaction';
import {
  constructEthersContractCaller,
  constructWeb3ContractCaller,
  constructAxiosFetcher,
  constructFetchFetcher,
  isFetcherError,
} from './helpers';
import {
  ConstructBaseInput,
  ConstructFetchInput,
  ConstructProviderFetchInput,
  Address,
} from './types';
import { UnionToIntersection } from 'ts-essentials';

export * from './constants';

// can import these individually
// and use in constructPartialSDK only what you need
// unused should tree-shake
export {
  constructApproveToken,
  constructGetBalances, // getAllowances is inside
  constructGetSpender, // this was getTokenTransferProxy
  constructGetTokens,
  constructBuildTx,
  constructEthersContractCaller,
  constructWeb3ContractCaller,
  constructAxiosFetcher,
  constructFetchFetcher,
  constructToken,
  constructGetAdapters,
  constructGetRate,
  isFetcherError,
};
// @TODO missing fro parity with older SDK:
// getRate, getRateByRoute

export type {
  ApproveTokenFunctions,
  GetBalancesFunctions,
  GetSpenderFunctions,
  GetTokensFunctions,
  BuildTxFunctions,
  ConstructFetchInput,
  ConstructProviderFetchInput,
  AdaptersFunctions as ConstructAdaptersFunctions,
  Token,
  Address,
  PriceString,
};

export type SDKConfig = ConstructProviderFetchInput & ConstructFetchInput;

export type AllSDKMethods = GetBalancesFunctions &
  GetTokensFunctions &
  GetSpenderFunctions &
  ApproveTokenFunctions &
  BuildTxFunctions &
  AdaptersFunctions &
  GetRateFunctions;

type AnyFunction = (...args: any[]) => any;

type SDKFunction<T extends ConstructBaseInput> = (
  config: T
) => Record<string, AnyFunction>;

type IntersectionOfReturns<Funcs extends AnyFunction[]> = UnionToIntersection<
  ReturnType<Funcs[number]>
>;

export const constructPartialSDK = <
  T extends ConstructBaseInput,
  Funcs extends [SDKFunction<T>, ...SDKFunction<T>[]]
>(
  config: T, // config is auto-inferred to cover the used functions
  ...funcs: Funcs
): IntersectionOfReturns<Funcs> => {
  const sdkFuncs = funcs.reduce<Partial<IntersectionOfReturns<Funcs>>>(
    (accum, func) => {
      const sdkSlice = func(config);
      return Object.assign(accum, sdkSlice);
    },
    {}
  );

  return sdkFuncs as IntersectionOfReturns<Funcs>;
};

export const constructSDK = (config: SDKConfig): AllSDKMethods =>
  // include all available functions
  constructPartialSDK(
    config,
    constructGetBalances,
    constructGetTokens,
    constructGetSpender,
    constructApproveToken,
    constructBuildTx,
    constructGetAdapters,
    constructGetRate
  );
