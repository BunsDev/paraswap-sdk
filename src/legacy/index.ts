import type { AxiosStatic } from 'axios';
import type Web3 from 'web3';
import type { SendOptions } from 'web3-eth-contract';
import type { ContractTransaction } from '@ethersproject/contracts';

import { API_URL, SwapSide } from '../constants';
import {
  AllSDKMethods,
  constructBuildTx,
  constructGetAdapters,
  constructGetBalances,
  constructGetSpender,
  constructGetTokens,
  constructPartialSDK,
  constructGetRate,
  constructFullSDK,
  PriceString,
  Address,
  OptimalRate,
} from '..';
import { assert } from 'ts-essentials';
import {
  constructAxiosFetcher,
  constructFetchFetcher,
  constructEthersContractCaller,
  constructWeb3ContractCaller,
  isFetcherError,
  Web3UnpromiEvent,
  EthersProviderDeps,
} from '../helpers';

import type { RateOptions } from '../methods/rates';
import type { BuildOptions, TransactionParams } from '../methods/transaction';
import type { AddressOrSymbol, Token, FetcherFunction } from '../types';
import type { Allowance } from '../methods/balance';
import { isDataWithError } from '../helpers/misc';

export type APIError = {
  message: string;
  status?: number;
  data?: any;
};
type Fetch = typeof fetch;

type TxResponse = Web3UnpromiEvent | ContractTransaction;

/** @deprecated */
export class ParaSwap {
  sdk: Partial<AllSDKMethods<TxResponse>> = {};
  fetcher: FetcherFunction;

  constructor(
    private network: number = 1,
    private apiURL: string = API_URL,
    public web3Provider?: Web3,
    public ethersDeps?: EthersProviderDeps, // need to be a provider with signer for approve requests
    public account?: Address,
    public axios?: AxiosStatic,
    public fetch?: Fetch
  ) {
    const fetcher = axios
      ? constructAxiosFetcher(axios)
      : fetch
      ? constructFetchFetcher(fetch)
      : null;

    assert(fetcher, 'at least one fetcher is needed');
    this.fetcher = fetcher;

    if (!web3Provider && !ethersDeps) {
      this.sdk = constructPartialSDK(
        { fetcher, apiURL, network },
        constructGetBalances,
        constructGetTokens,
        constructGetSpender,
        constructBuildTx,
        constructGetAdapters,
        constructGetRate
      );

      return;
    }

    const contractCaller = ethersDeps
      ? constructEthersContractCaller(ethersDeps, account)
      : web3Provider
      ? constructWeb3ContractCaller(web3Provider, account)
      : null;

    if (contractCaller) {
      this.sdk = constructFullSDK<TxResponse>({
        fetcher,
        contractCaller,
        apiURL,
        network,
      });
    }
  }

  private static handleAPIError(e: unknown): APIError {
    // @CONSIDER if some errors should not be replaced
    if (!isFetcherError(e)) {
      return { message: `Unknown error: ${e}` };
    }

    if (!e.response) {
      return { message: e.message };
    }

    const { status, data } = e.response;

    return {
      status,
      message: isDataWithError(data) ? data.error : e.message,
      data,
    };
  }

  private static async extractHasFromTxResponse(
    txResponse: TxResponse
  ): Promise<string> {
    if ('once' in txResponse) {
      return new Promise<string>((resolve, reject) => {
        txResponse.once('transactionHash', resolve);
        txResponse.once('error', reject);
      });
    }

    const { hash } = await txResponse;
    return hash;
  }

  setWeb3Provider(web3Provider: Web3, account?: string): this {
    const contractCaller = constructWeb3ContractCaller(web3Provider, account);
    const { apiURL, network, fetcher } = this;

    this.sdk = constructFullSDK({
      fetcher,
      contractCaller,
      apiURL,
      network,
    });

    this.web3Provider = web3Provider;
    this.ethersDeps = undefined;
    this.account = account;

    return this;
  }

  setEthersProvider(ethersDeps: EthersProviderDeps, account?: string): this {
    const contractCaller = constructEthersContractCaller(ethersDeps, account);
    const { apiURL, network, fetcher } = this;

    this.sdk = constructFullSDK({
      fetcher,
      contractCaller,
      apiURL,
      network,
    });

    this.web3Provider = undefined;
    this.ethersDeps = ethersDeps;
    this.account = account;

    return this;
  }

  // @CONSIDER I still think there's no need for a class Token
  async getTokens(): Promise<Token[] | APIError> {
    assert(this.sdk.getTokens, 'sdk must be initialized with a fetcher');
    try {
      return await this.sdk.getTokens();
    } catch (e) {
      return ParaSwap.handleAPIError(e);
    }
  }

  async getAdapters() {
    assert(this.sdk.getAdapters, 'sdk must be initialized with a fetcher');
    try {
      return await this.sdk.getAdapters({ type: 'object' });
    } catch (e) {
      return ParaSwap.handleAPIError(e);
    }
  }

  async getRateByRoute(
    route: AddressOrSymbol[],
    amount: PriceString,
    userAddress?: Address,
    side: SwapSide = SwapSide.SELL,
    options?: RateOptions,
    srcDecimals?: number,
    destDecimals?: number
  ): Promise<OptimalRate | APIError> {
    assert(this.sdk.getRateByRoute, 'sdk must be initialized with a fetcher');
    if (route.length < 2) {
      return { message: 'Invalid Route' };
    }

    try {
      return await this.sdk.getRateByRoute({
        route,
        amount,
        userAddress,
        side,
        options,
        srcDecimals,
        destDecimals,
      });
    } catch (e) {
      // @TODO this overrides any non FetchError,
      // including Error('Invalid DEX list')
      return ParaSwap.handleAPIError(e);
    }
  }

  async getRate(
    srcToken: AddressOrSymbol,
    destToken: AddressOrSymbol,
    amount: PriceString,
    userAddress?: Address,
    side: SwapSide = SwapSide.SELL,
    options: RateOptions = {},
    srcDecimals?: number,
    destDecimals?: number
  ): Promise<OptimalRate | APIError> {
    assert(this.sdk.getRate, 'sdk must be initialized with a fetcher');
    try {
      return await this.sdk.getRate({
        srcToken,
        destToken,
        amount,
        userAddress,
        side,
        options,
        srcDecimals,
        destDecimals,
      });
    } catch (e) {
      return ParaSwap.handleAPIError(e);
    }
  }

  async buildTx(
    srcToken: Address,
    destToken: Address,
    srcAmount: PriceString,
    destAmount: PriceString,
    priceRoute: OptimalRate,
    userAddress: Address,
    partner?: string,
    partnerAddress?: string,
    partnerFeeBps?: number,
    receiver?: Address,
    options: BuildOptions = {},
    srcDecimals?: number,
    destDecimals?: number,
    permit?: string,
    deadline?: string
  ): Promise<TransactionParams | APIError> {
    assert(this.sdk.buildTx, 'sdk must be initialized with a fetcher');
    try {
      return await this.sdk.buildTx(
        {
          srcToken,
          destToken,
          srcAmount,
          destAmount,
          priceRoute,
          userAddress,
          partner,
          partnerAddress,
          partnerFeeBps,
          receiver,
          srcDecimals,
          destDecimals,
          permit,
          deadline,
        },
        options
      );
    } catch (e) {
      return ParaSwap.handleAPIError(e);
    }
  }

  async getTokenTransferProxy(_provider?: any): Promise<Address | APIError> {
    assert(this.sdk.getSpender, 'sdk must be initialized with a fetcher');
    try {
      return await this.sdk.getSpender();
    } catch (e) {
      return ParaSwap.handleAPIError(e);
    }
  }

  async getAllowances(
    userAddress: Address,
    tokenAddresses: Address[]
  ): Promise<Allowance[] | APIError> {
    assert(this.sdk.getAllowances, 'sdk must be initialized with a fetcher');
    try {
      return await this.sdk.getAllowances(userAddress, tokenAddresses);
    } catch (e) {
      return ParaSwap.handleAPIError(e);
    }
  }

  async getAllowance(
    userAddress: Address,
    tokenAddress: Address
  ): Promise<Allowance | APIError> {
    assert(this.sdk.getAllowance, 'sdk must be initialized with a fetcher');
    try {
      return await this.sdk.getAllowance(userAddress, tokenAddress);
    } catch (e) {
      return ParaSwap.handleAPIError(e);
    }
  }

  async approveTokenBulk(
    amount: PriceString,
    userAddress: Address,
    tokenAddresses: Address[],
    /** @deprecated */
    _provider?: any // not used, can't detect if Ethers or Web3 provider without importing them
  ): Promise<string[] | APIError> {
    // @TODO expand sendOptions
    assert(
      this.sdk.approveTokenBulk,
      'sdk must be initialized with a provider'
    );
    try {
      // @TODO allow to pass Web3 specific sendOptions ({from: userAddress})
      const txResponses = await this.sdk.approveTokenBulk(
        amount,
        tokenAddresses
      );

      return await Promise.all(
        txResponses.map(ParaSwap.extractHasFromTxResponse)
      );
    } catch (e) {
      return ParaSwap.handleAPIError(e);
    }
  }

  async approveToken(
    amount: PriceString,
    userAddress: Address,
    tokenAddress: Address,
    /** @deprecated */
    _provider?: any, // not used, can't detect if Ethers or Web3 provider without importing them
    sendOptions?: Omit<SendOptions, 'from'>
  ): Promise<string | APIError> {
    // @TODO expand sendOptions
    assert(this.sdk.approveToken, 'sdk must be initialized with a provider');
    try {
      // @TODO allow to pass Web3 specific sendOptions ({from: userAddress})
      const txResponse = await this.sdk.approveToken(amount, tokenAddress);

      return await ParaSwap.extractHasFromTxResponse(txResponse);
    } catch (e) {
      return ParaSwap.handleAPIError(e);
    }
  }

  async getMarketNames(): Promise<string[] | APIError> {
    assert(this.sdk.getAdapters, 'sdk must be initialized with a fetcher');
    try {
      return await this.sdk.getAdapters({ type: 'list', namesOnly: true });
    } catch (e) {
      return ParaSwap.handleAPIError(e);
    }
  }

  async getBalance(
    userAddress: Address,
    token: AddressOrSymbol
  ): Promise<Token | APIError> {
    assert(this.sdk.getBalance, 'sdk must be initialized with a fetcher');
    try {
      return await this.sdk.getBalance(userAddress, token);
    } catch (e) {
      return ParaSwap.handleAPIError(e);
    }
  }

  async getBalances(userAddress: Address): Promise<Token[] | APIError> {
    assert(this.sdk.getBalances, 'sdk must be initialized with a fetcher');
    try {
      return await this.sdk.getBalances(userAddress);
    } catch (e) {
      return ParaSwap.handleAPIError(e);
    }
  }
}
