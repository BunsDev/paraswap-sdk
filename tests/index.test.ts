import * as dotenv from 'dotenv';
import Web3 from 'web3';
import { ethers } from 'ethers';
import axios from 'axios';
import fetch from 'isomorphic-unfetch';
import {
  AllSDKMethods,
  constructAxiosFetcher,
  constructEthersContractCaller,
  constructFetchFetcher,
  constructSDK,
} from '../src';
import BigNumber from 'bignumber.js';
import { SwapSide } from '../src';
import { Adapters } from 'paraswap-core';
import { Allowance } from '../src/balance';
import { APIError } from '../src/legacy';
import erc20abi from './abi/ERC20.json';

import ganache from 'ganache';

dotenv.config();

jest.setTimeout(30 * 1000);

declare let process: any;

const ETH = '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee';
const DAI = '0x6B175474E89094C44Da98b954EedeAC495271d0F';
const BAT = '0x0d8775f648430679a709e98d2b0cb6250d2887ef';
const MANA = '0x0f5d2fb29fb7d3cfee444a200298f468908cc942';

const TESTING_ENV = true;
const PROVIDER_URL = process.env.PROVIDER_URL;
const network = 1;
const srcToken = ETH;
const destToken = DAI;
const srcAmount = (1 * 1e18).toString(); //The source amount multiplied by its decimals

const referrer = 'sdk-test';

const wallet = ethers.Wallet.createRandom();

const ganacheProvider = ganache.provider({
  wallet: {
    accounts: [{ balance: 8e18, secretKey: wallet.privateKey }],
  },
  fork: {
    url: PROVIDER_URL,
  },
  chain: {
    chainId: 1,
  },
});

const web3provider = new Web3(ganacheProvider as any);

const ethersProvider = new ethers.providers.Web3Provider(
  ganacheProvider as any
);

const fetchFetcher = constructFetchFetcher(fetch);
const axiosFetcher = constructAxiosFetcher(axios);

const signer = wallet.connect(ethersProvider);
const senderAddress = signer.address;

const ethersContractCaller = constructEthersContractCaller(
  {
    providerOrSigner: signer,
    Contract: ethers.Contract,
  },
  senderAddress
);

const web3ContractCaller = constructEthersContractCaller(
  {
    providerOrSigner: signer,
    Contract: ethers.Contract,
  },
  senderAddress
);

describe.each([
  ['fetchFetcher & web3ContractCaller', fetchFetcher, web3ContractCaller],
  ['axiosFetcher & ethersContractCaller', axiosFetcher, ethersContractCaller],
])('ParaSwap SDK: %s', (testName, fetcher, contractCaller) => {
  let paraSwap: AllSDKMethods;

  beforeAll(async () => {
    paraSwap = constructSDK({
      network,
      fetcher,
      contractCaller,
    });
    // @ts-expect-error
    paraSwap.adapters = (await paraSwap.getAdapters()) as Adapters;
  });

  test('Get_Balance', async () => {
    const balance = await paraSwap.getBalance(senderAddress, ETH);
    expect(balance).toBeDefined();
  });

  test('Get_Markets', async () => {
    const markets = await paraSwap.getAdapters({
      type: 'list',
      namesOnly: true,
    });
    expect((markets as string[]).length).toBeGreaterThan(15);
  });

  test('Get_Tokens', async () => {
    const tokens = await paraSwap.getTokens();

    expect(Array.isArray(tokens)).toBe(true);
    expect(tokens.length).toBeGreaterThan(0);
    expect(tokens[0]).toEqual(
      expect.objectContaining({
        symbol: expect.any(String),
        address: expect.any(String),
        decimals: expect.any(Number),
      })
    );
  });

  test('Get_Rates', async () => {
    const priceRoute = await paraSwap.getRate({
      srcToken: ETH,
      destToken: DAI,
      amount: srcAmount,
      userAddress: senderAddress,
      side: SwapSide.SELL,
      options: {
        includeDEXS: 'UniswapV2',
        otherExchangePrices: true,
      },
    });

    const { destAmount, bestRoute, others } = priceRoute;

    expect(typeof destAmount).toBe('string');

    expect(Array.isArray(bestRoute)).toBe(true);

    const swapExchange = bestRoute[0]?.swaps[0]?.swapExchanges[0];

    expect(swapExchange).toBeDefined();

    expect(typeof swapExchange.destAmount).toBe('string');
    expect(new BigNumber(swapExchange.destAmount).isNaN()).toBe(false);

    expect(typeof swapExchange.exchange).toBe('string');

    expect(typeof bestRoute[0].percent).toBe('number');
    expect(new BigNumber(bestRoute[0].percent).isNaN()).toBe(false);

    expect(typeof swapExchange.srcAmount).toBe('string');
    expect(new BigNumber(swapExchange.srcAmount).isNaN()).toBe(false);

    expect(Array.isArray(others)).toBe(true);

    expect(typeof others![0].exchange).toBe('string');

    expect(typeof others![0].unit).toBe('string');
    expect(new BigNumber(others![0].unit as string).isNaN()).toBe(false);
  });

  test('Get_Spender', async () => {
    const spender = await paraSwap.getSpender();
    expect(web3provider.utils.isAddress(spender as string));
  });

  test('Get_Allowance', async () => {
    const allowance = await paraSwap.getAllowance(senderAddress, DAI);

    if (!allowance || (allowance as APIError).message) {
      return;
    }

    expect(new BigNumber((allowance as Allowance).allowance).isNaN()).toBe(
      false
    );
  });

  test('Get_Allowances', async () => {
    const allowances = await paraSwap.getAllowances(senderAddress, [
      DAI,
      BAT,
      MANA,
    ]);

    await Promise.all(
      allowances.map((allowance) =>
        expect(new BigNumber(allowance.allowance).isNaN()).toBe(false)
      )
    );
  });

  test('Get_Adapters', async () => {
    const adapters = await paraSwap.getAdapters({ type: 'object' });
    expect(adapters.paraswappool[0].adapter).toBeDefined();
    expect(adapters.uniswapv2[0].adapter).toBeDefined();
    expect(adapters.uniswapv2[0].index).toBeDefined();
    expect(adapters.kyber[0].adapter).toBeDefined();
    expect(adapters.kyber[0].index).toBeDefined();
  });

  test('Build_Tx', async () => {
    const destToken = '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48';
    const priceRoute = await paraSwap.getRate({
      srcToken,
      destToken,
      amount: srcAmount,
      userAddress: senderAddress,
      side: SwapSide.SELL,
      options: {
        includeDEXS: 'UniswapV2',
      },
    });

    const destAmount = new BigNumber(priceRoute.destAmount)
      .times(0.99)
      .toFixed(0);

    const txParams = await paraSwap.buildTx(
      {
        srcToken,
        destToken,
        srcAmount,
        destAmount,
        priceRoute,
        userAddress: senderAddress,
        partner: referrer,
      },
      { ignoreChecks: true }
    );

    expect(typeof txParams).toBe('object');
  });
  if (TESTING_ENV) {
    test('Build_tx_legacy', async () => {
      const priceRoute = await paraSwap.getRate({
        srcToken,
        destToken,
        amount: srcAmount,
        userAddress: senderAddress,
        side: SwapSide.SELL,
        options: {
          includeDEXS: 'Uniswap,UniswapV2,Balancer,Oasis',
        },
      });
      const destAmount = new BigNumber(priceRoute.destAmount)
        .times(0.99)
        .toFixed(0);

      const transaction = await paraSwap.buildTx(
        {
          srcToken,
          destToken,
          srcAmount,
          destAmount,
          priceRoute,
          userAddress: senderAddress,
          partner: referrer,
        },
        { ignoreChecks: true }
      );
      expect(typeof transaction).toBe('object');
    });
    test('Build_and_Send_Tx', async () => {
      const priceRoute = await paraSwap.getRate({
        srcToken,
        destToken,
        amount: srcAmount,
        userAddress: senderAddress,
        side: SwapSide.SELL,
        options: {
          includeDEXS: 'Uniswap,UniswapV2,Balancer,Oasis',
        },
      });

      const destAmount = new BigNumber(priceRoute.destAmount)
        .times(0.99)
        .toFixed(0);

      const txParams = await paraSwap.buildTx(
        {
          srcToken,
          destToken,
          srcAmount,
          destAmount,
          priceRoute,
          userAddress: signer.address,
          partner: referrer,
        },
        { ignoreChecks: true }
      );

      const transaction = {
        ...txParams,
        gasPrice: '0x' + new BigNumber(txParams.gasPrice).toString(16),
        gasLimit: '0x' + new BigNumber(5000000).toString(16),
        value: '0x' + new BigNumber(txParams.value).toString(16),
      };
      const toContract = new ethers.Contract(
        destToken,
        erc20abi,
        ethersProvider
      );
      const beforeFromBalance = await ethersProvider.getBalance(signer.address);
      const beforeToBalance = await toContract.balanceOf(signer.address);

      const txr = await signer.sendTransaction(transaction);
      await txr.wait(1);
      const afterFromBalance = await ethersProvider.getBalance(signer.address);
      const afterToBalance = await toContract.balanceOf(signer.address);
      expect(beforeFromBalance.gt(afterFromBalance)).toBeTruthy();
      expect(beforeToBalance.lt(afterToBalance)).toBeTruthy();
    });
    test('Build_and_Send_Tx_BUY', async () => {
      const destAmount = srcAmount;
      const priceRoute = await paraSwap.getRate({
        srcToken,
        destToken,
        amount: destAmount,
        userAddress: senderAddress,
        side: SwapSide.BUY,
        options: { includeDEXS: 'Uniswap,UniswapV2,Balancer,Oasis' },
      });
      const _srcAmount = new BigNumber(priceRoute.srcAmount)
        .times(1.1)
        .toFixed(0);

      const txParams = await paraSwap.buildTx(
        {
          srcToken,
          destToken,
          srcAmount: _srcAmount,
          destAmount,
          priceRoute,
          userAddress: signer.address,
          partner: referrer,
        },
        { ignoreChecks: true }
      );

      const transaction = {
        ...txParams,
        gasPrice: '0x' + new BigNumber(txParams.gasPrice).toString(16),
        gasLimit: '0x' + new BigNumber(5000000).toString(16),
        value: '0x' + new BigNumber(txParams.value).toString(16),
      };
      const toContract = new ethers.Contract(
        destToken,
        erc20abi,
        ethersProvider
      );
      const beforeFromBalance = await ethersProvider.getBalance(signer.address);
      const beforeToBalance = await toContract.balanceOf(signer.address);

      const txr = await signer.sendTransaction(transaction);
      await txr.wait(1);
      const afterFromBalance = await ethersProvider.getBalance(signer.address);
      const afterToBalance = await toContract.balanceOf(signer.address);
      expect(beforeFromBalance.gt(afterFromBalance)).toBeTruthy();
      expect(beforeToBalance.lt(afterToBalance)).toBeTruthy();
    });
  }
});
