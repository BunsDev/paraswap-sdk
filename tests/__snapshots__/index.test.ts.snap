// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParaSwap SDK: fetching methods: axiosFetcher Get_Contracts: axiosFetcher 1`] = `
Object {
  "AugustusSwapper": "0xDEF171Fe48CF0115B1d80b88dc8eAB59176FEe57",
  "TokenTransferProxy": "0x216b4b4ba9f3e719726886d34a177484278bfcae",
}
`;

exports[`ParaSwap SDK: fetching methods: fetchFetcher Get_Contracts: fetchFetcher 1`] = `
Object {
  "AugustusSwapper": "0xDEF171Fe48CF0115B1d80b88dc8eAB59176FEe57",
  "TokenTransferProxy": "0x216b4b4ba9f3e719726886d34a177484278bfcae",
}
`;
